albatross/core/prediction.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_CORE_PREDICTION_H
+#define ALBATROSS_CORE_PREDICTION_H
+
+namespace albatross {
+
+// This is effectively just a container that allows us to develop methods
+// which behave different conditional on the type of predictions desired.
+template <typename T> struct PredictTypeIdentity { typedef T type; };
+
+template <typename ModelType, typename FeatureType, typename FitType>
+class Prediction {
+
+public:
+  Prediction(const FitModel<ModelType, FitType> &fit_model,
+             const std::vector<FeatureType> &features)
+      : fit_model_(fit_model), features_(features) {}
+
+  /*
+   * MEAN
+   */
+  template <
```

Comment:
These methods define all the logic about things like "If a `predict` method exists for some `Distribution<T>` but not for the mean (`Eigen::VectorXd`) we can get the mean from `Distribution<T>.mean`.  Typically the complexity of producing a full `JointDistribution` is larger than a `MarginalDistribution` which is larger than the mean only so these are designed to define `mean`, `marginal` and `joint` using the least computationally complex method available.